service/matching/taskListManager.go

```diff
@@ -573,6 +566,42 @@ func (c *taskListManagerImpl) shouldReload() bool {
 	return c.config.EnableTasklistIsolation() != c.enableIsolation && c.taskListKind != types.TaskListKindSticky
 }
 
+func (c *taskListManagerImpl) getIsolationGroupForTask(ctx context.Context, taskInfo *persistence.TaskInfo) string {
+	if c.isIsolationEnabled() && len(taskInfo.PartitionConfig) > 0 {
+		partitionConfig := make(map[string]string)
+		for k, v := range taskInfo.PartitionConfig {
+			partitionConfig[k] = v
+		}
+		partitionConfig[partition.WorkflowRunIDKey] = taskInfo.RunID
+		var isolationGroup string
+		group, err := c.partitioner.GetIsolationGroupByDomainID(ctx, taskInfo.DomainID, partitionConfig)
+		if err != nil {
+			c.logger.Error("Failed to get isolation group from partition library", tag.WorkflowID(taskInfo.WorkflowID), tag.WorkflowRunID(taskInfo.RunID), tag.TaskID(taskInfo.TaskID), tag.Error(err))
+		} else {
+			isolationGroup = group
+		}
+		// check if the isolation group has pollers, if not use the fallback logic
+		pollerIsolationGroups := c.pollerHistory.getPollerIsolaionGroups(time.Now().Add(-time.Minute))
+		if len(pollerIsolationGroups) == 0 {
+			// we don't have any pollers, use the original isolation group and wait for pollers' arriving
+			return isolationGroup
+		}
+		if isolationGroup != "" {
+			for _, pollerGroup := range pollerIsolationGroups {
+				if pollerGroup == isolationGroup {
+					return isolationGroup
+				}
+			}
+			c.logger.Warn("Failed to find a poller from the given isolation group", tag.WorkflowID(taskInfo.WorkflowID), tag.WorkflowRunID(taskInfo.RunID), tag.TaskID(taskInfo.TaskID))
+		}
+		// if we hit this branch, either we failed to get isolation group from the partition library, or we don't have pollers from the given isolation group
+		// use runID to deterministically choose an isolation group with pollers
+		hashv := farm.Hash32([]byte(taskInfo.RunID))
```

Comment:
I do think we should pull this into either the partition library or otherwise ensure there's a single place where we're doing partitioning. Spreading the logic around is I think is risky when you accidentally change one but not the other.